server/svix-server/src/core/message_app.rs

```diff
@@ -0,0 +1,157 @@
+    error::{Error, Result},
+};
+
+/// The information cached during the creation of a message. Includes a [`Vec`] of all endpoints
+/// associated with the given application and organization ID.
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub struct CreateMessageApp {
+    pub id: ApplicationId,
+    pub uid: Option<ApplicationUid>,
+    pub org_id: OrganizationId,
+    pub rate_limit: Option<u16>,
+    pub endpoints: Vec<CreateMessageEndpoint>,
+    pub deleted: bool,
+}
+
+impl CreateMessageApp {
+    /// Fetch all requisite information for creating a [`CreateMessageApp`] from the PostgreSQL
+    /// database
+    async fn fetch_from_pg_by_model(
+        db: &DatabaseTransaction,
+        app: application::Model,
+    ) -> Result<CreateMessageApp> {
+        let endpoints = endpoint::Entity::secure_find(app.id.clone())
+            .all(db)
+            .await?
+            .into_iter()
+            .map(TryInto::try_into)
+            .collect::<Result<Vec<_>>>()?;
+
+        Ok(CreateMessageApp {
+            id: app.id,
+            uid: app.uid,
+            org_id: app.org_id,
+            rate_limit: app
+                .rate_limit
+                .map(|v| v.try_into())
+                .transpose()
+                .map_err(|_| {
+                    Error::Validation("Application rate limit out of bounds".to_owned())
+                })?,
+            endpoints,
+            deleted: app.deleted,
+        })
+    }
+
+    /// Fetches all information for creating a [`CreateMessageApp`] from the Redis cache if it
+    /// exists or from PostgreSQL otherwise. If the RedisCache is Some, but does not contain the
+    /// requisite information, fetch it from PostgreSQL and insert the data into the cache.
+    pub async fn layered_fetch(
+        redis: Option<&RedisCache>,
+        pg: &DatabaseConnection,
+        app: Option<application::Model>,
+        app_id: ApplicationId,
+        org_id: OrganizationId,
+    ) -> Result<Option<CreateMessageApp>> {
+        let cache_key = AppEndpointKey::new(org_id.clone(), app_id.clone());
+
+        // First check Redis
+        if let Some(redis) = redis {
+            if let Ok(Some(cma)) = redis.get(&cache_key).await {
+                return Ok(Some(cma));
+            }
+        }
+
+        // Then check PostgreSQL
+        let db = pg.begin().await?;
+        // Fetch the [`application::Model`] either given or from the ID
+        let app = if let Some(app) = app {
+            app
+        } else if let Some(app) = application::Entity::secure_find_by_id(org_id, app_id)
+            .one(&db)
+            .await?
+        {
+            app
+        } else {
+            return Ok(None);
+        };
+
+        // Fetch the actual [`CreateMessageApp`]
+        let out = Self::fetch_from_pg_by_model(&db, app).await?;
+
+        // Insert it into Redis
+        if let Some(redis) = redis {
+            let _ = redis.set(&cache_key, &out, Duration::from_secs(30)).await;
```

Comment:
Can you move this duration into a constant, or even just a variable? We'll want to make it configurable in the near future, and anyway we want to have it be set somewhere explicit.